sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ToString.java

```diff
@@ -18,51 +18,176 @@
  * {@link SerializableFunction} using {@link MapElements#via(SerializableFunction)}
  */
 public final class ToString {
+  private ToString() {
+    // do not instantiate
+  }
 
   /**
    * Returns a {@code PTransform<PCollection, PCollection<String>>} which transforms each
    * element of the input {@link PCollection} to a {@link String} using the
    * {@link Object#toString} method.
    */
-  public static PTransform<PCollection<?>, PCollection<String>> element() {
-    return new Default();
+  public static PTransform<PCollection<?>, PCollection<String>> of() {
+    return new SimpleToString();
   }
 
-  private ToString() {
+  /**
+   * Returns a {@code PTransform<PCollection, PCollection<String>>} which transforms each
+   * element of the input {@link PCollection} to a {@link String} using the
+   * {@link Object#toString} method.
+   */
+  public static PTransform<PCollection<? extends KV<?, ?>>, PCollection<String>> kv() {
+    return kv(",");
+  }
+
+  /**
+   * Returns a {@code PTransform<PCollection, PCollection<String>>} which transforms each
+   * element of the input {@link PCollection} to a {@link String} using the
+   * {@link Object#toString} method.
+   * @param delimiter The delimiter to put between the key and value
+   */
+  public static PTransform<PCollection<? extends KV<?, ?>>,
+          PCollection<String>> kv(String delimiter) {
+    return new KVToString(delimiter);
+  }
+
+  /**
+   * Returns a {@code PTransform<PCollection, PCollection<String>>} which transforms each
+   * element of the input {@link PCollection} to a {@link String} using the
+   * {@link Object#toString} method.
+   */
+  public static PTransform<PCollection<? extends Iterable<?>>, PCollection<String>> iterable() {
+    return iterable(",");
+  }
+
+  /**
+   * Returns a {@code PTransform<PCollection, PCollection<String>>} which transforms each
+   * element of the input {@link PCollection} to a {@link String} using the
+   * {@link Object#toString} method.
+   * @param delimiter The delimiter to put between the key and value
+   */
+  public static PTransform<PCollection<? extends Iterable<?>>,
+          PCollection<String>> iterable(String delimiter) {
+    return new IterablesToString(delimiter);
   }
 
   /**
    * A {@link PTransform} that converts a {@code PCollection} to a {@code PCollection<String>}
    * using the {@link  Object#toString} method.
+   *
+   * <p>Example of use:
+   * <pre> {@code
+   * PCollection<Long> longs = ...;
+   * PCollection<String> strings = longs.apply(ToString.of());
+   * } </pre>
+   *
+   *
+   * <p><b>Note</b>: For any custom string conversion and formatting, we recommend applying your own
+   * {@link SerializableFunction} using {@link MapElements#via(SerializableFunction)}
    */
-  private static final class Default extends PTransform<PCollection<?>, PCollection<String>> {
+  public static final class SimpleToString extends PTransform<PCollection<?>, PCollection<String>> {
```

Comment:
As far as I know, the SDK is consistent, or fairly close. But there may be a hidden underlying principle that makes it look inconsistent otherwise: least visibility. Make nothing "more public" than it needs to be.

* If the user needs to know a specific return type, we let them. Usually, this means because they want to further configure the return value. For example, `ParDo.of(doFn)` does not return just `PTransform` because you still want to configure the result by, say, adding side inputs.

* If the user does not need a specific return type, return the most general thing you can. We can always make the return type more specific later -- I believe that to be a backwards-compatible change.